From 99f31474b9a024f3b4e7ae7214baba467475e254 Mon Sep 17 00:00:00 2001
From: Chelsio Cudbg <chelsiocudbg@users.noreply.github.com>
Date: Wed, 10 Jun 2026 08:26:25 -0400
Subject: [PATCH 9/9] iw_cxgb4: fix rdma bind failure

when iommu is enabled mapping is not happeing properly,
fix this issue by passing proper addresses.
---
 drivers/infiniband/hw/cxgb4/cq.c     |  2 +-
 drivers/infiniband/hw/cxgb4/device.c | 12 ++++++------
 drivers/infiniband/hw/cxgb4/qp.c     |  8 ++++----
 3 files changed, 11 insertions(+), 11 deletions(-)

diff --git a/drivers/infiniband/hw/cxgb4/cq.c b/drivers/infiniband/hw/cxgb4/cq.c
index 1828ab3507bc..ff56743cde57 100644
--- a/drivers/infiniband/hw/cxgb4/cq.c
+++ b/drivers/infiniband/hw/cxgb4/cq.c
@@ -1499,7 +1499,7 @@ int c4iw_create_cq(struct ib_cq *ibcq, const struct ib_cq_init_attr *attr,
 			goto err_free_mm2;
 
 		mm->key = uresp.key;
-		mm->addr = virt_to_phys(chp->cq.queue);
+		mm->addr = 0;
 		mm->vaddr = chp->cq.queue;
 		mm->dma_addr = chp->cq.dma_addr;
 		mm->len = chp->cq.memsize;
diff --git a/drivers/infiniband/hw/cxgb4/device.c b/drivers/infiniband/hw/cxgb4/device.c
index e6bd679ee1a2..0eccbc1caf39 100644
--- a/drivers/infiniband/hw/cxgb4/device.c
+++ b/drivers/infiniband/hw/cxgb4/device.c
@@ -954,8 +954,8 @@ static int c4iw_rdev_open(struct c4iw_rdev *rdev)
 		goto err_destroy_pblpool;
 	}
 
-	rdev->status_page = dma_alloc_coherent(rdev->lldi.dev, PAGE_SIZE,
-			&rdev->daddr, GFP_KERNEL);
+	rdev->status_page = (struct t4_dev_status_page *)
+			     __get_free_page(GFP_KERNEL);
 	if (!rdev->status_page) {
 		pr_err("error allocating status page\n");
 		goto err_destroy_rrqtpool;
@@ -997,8 +997,9 @@ static int c4iw_rdev_open(struct c4iw_rdev *rdev)
 
 	return 0;
 err_free_status_page:
-	dma_free_coherent(rdev->lldi.dev, PAGE_SIZE, rdev->status_page,
-			rdev->daddr);
+	if (c4iw_wr_log && rdev->wr_log)
+		kfree(rdev->wr_log);
+	free_page((unsigned long)rdev->status_page);
 err_destroy_rrqtpool:
 	c4iw_rrqtpool_destroy(rdev);
 err_destroy_pblpool:
@@ -1015,8 +1016,7 @@ static void c4iw_rdev_close(struct c4iw_rdev *rdev)
 {
 	if (rdev->wr_log)
 		kfree(rdev->wr_log);
-	dma_free_coherent(rdev->lldi.dev, PAGE_SIZE, rdev->status_page,
-			rdev->daddr);
+	free_page((unsigned long)rdev->status_page);
 	kfree(rdev->fids);
 	c4iw_pblpool_destroy(rdev);
 	c4iw_rqtpool_destroy(rdev);
diff --git a/drivers/infiniband/hw/cxgb4/qp.c b/drivers/infiniband/hw/cxgb4/qp.c
index 4c9cf7ebf12f..e65237b288ac 100644
--- a/drivers/infiniband/hw/cxgb4/qp.c
+++ b/drivers/infiniband/hw/cxgb4/qp.c
@@ -5229,7 +5229,7 @@ static int create_rc_qp(struct ib_qp *qp, struct ib_qp_init_attr *attrs,
 		if (ret)
 			goto err_free_ma_sync_key;
 		sq_key_mm->key = uresp.sq_key;
-		sq_key_mm->addr = qhp->wq.sq.phys_addr;
+		sq_key_mm->addr = 0;
 		sq_key_mm->vaddr = qhp->wq.sq.queue;
 		sq_key_mm->dma_addr = qhp->wq.sq.dma_addr;
 		sq_key_mm->len = PAGE_ALIGN(qhp->wq.sq.memsize);
@@ -5237,7 +5237,7 @@ static int create_rc_qp(struct ib_qp *qp, struct ib_qp_init_attr *attrs,
 		insert_mmap(ucontext, sq_key_mm);
 		if (!attrs->srq) {
 			rq_key_mm->key = uresp.rq_key;
-			rq_key_mm->addr = virt_to_phys(qhp->wq.rq.queue);
+			rq_key_mm->addr = 0;
 			rq_key_mm->vaddr = qhp->wq.rq.queue;
 			rq_key_mm->dma_addr = qhp->wq.rq.dma_addr;
 			rq_key_mm->len = PAGE_ALIGN(qhp->wq.rq.memsize);
@@ -5245,7 +5245,7 @@ static int create_rc_qp(struct ib_qp *qp, struct ib_qp_init_attr *attrs,
 			insert_mmap(ucontext, rq_key_mm);
 		}
 		sq_db_key_mm->key = uresp.sq_db_gts_key;
-		sq_db_key_mm->addr = qhp->wq.sq.db_pa;
+		sq_db_key_mm->addr = (u64)(unsigned long)qhp->wq.sq.db_pa;
 		sq_db_key_mm->vaddr = NULL;
 		sq_db_key_mm->dma_addr = 0;
 		sq_db_key_mm->len = PAGE_SIZE;
@@ -5253,7 +5253,7 @@ static int create_rc_qp(struct ib_qp *qp, struct ib_qp_init_attr *attrs,
 		insert_mmap(ucontext, sq_db_key_mm);
 		if (!attrs->srq) {
 			rq_db_key_mm->key = uresp.rq_db_gts_key;
-			rq_db_key_mm->addr = qhp->wq.rq.db_pa;
+			rq_db_key_mm->addr = (u64)(unsigned long)qhp->wq.rq.db_pa;
 			rq_db_key_mm->len = PAGE_SIZE;
 			rq_db_key_mm->vaddr = NULL;
 			rq_db_key_mm->dma_addr = 0;
-- 
2.39.1

